README.md
Author: igor.zavoychinskiy@gmail.com

GitHub: https://github.com/ihsoft/KSPDev_ReleaseBuilder

Read discussions, ask questions and suggest features on
[forum](http://forum.kerbalspaceprogram.com/index.php?/topic/150786-12-kspdev-logconsole-utils).

# KSPDev: Kerbal Development tools - Release Tools

It's a set of tools for making and publishing the KSP mod releases.

* [`ReleaseBuilder`](https://github.com/ihsoft/KSPDev_ReleaseBuilder/blob/master/KspReleaseBuilder.py)
  is flexible script that allows building mod releases for KSP quick and easy. Just define mod's
  structure and the script will manage everything else: from updating MinAVC version file to
  preparing final ZIP package.
  See [Wiki](https://github.com/ihsoft/KSPDev_ReleaseBuilder/wiki/Release-builder-script)
  for more details.
* [`PublishGitHub`](https://github.com/ihsoft/KSPDev_ReleaseBuilder/blob/master/PublishGitHub.py)
  is a script that allows creating a release on GutHub with the latest release archive and
  changelog description. The release can be created as _DRAFT_ to allow extra steps to be done
  before actually publishing the release to the world.
  See [Wiki](https://github.com/ihsoft/KSPDev_ReleaseBuilder/wiki/Release-publishing-tools#github)
  for more details.
* [`PublishSpacedock`](https://github.com/ihsoft/KSPDev_ReleaseBuilder/blob/master/PublishSpacedock.py)
  is a script that uplaods a mod version to _Spacedock_ with all the properties  set and the
  version details populated.
  See [Wiki](https://github.com/ihsoft/KSPDev_ReleaseBuilder/wiki/Release-publishing-tools#spacedock)
  for more details.
* [`PublishCurseForge`](https://github.com/ihsoft/KSPDev_ReleaseBuilder/blob/master/PublishCurseForge.py)
  is a script that uplaods a mod version to _CurseForge_ with all the properties set and the
  version details populated.
  See [Wiki](https://github.com/ihsoft/KSPDev_ReleaseBuilder/wiki/Release-publishing-tools#curseforge)
  for more details.

# Installation

All the scripts need Python 2.7. The GitHub script needs a minimum version `2.7.11`. However,
it's always the best to upgrade to the
[latest 2.7 version available](https://www.python.org/downloads/).


CHANGELOG.md
# KPSPDev: ReleaseTools v1.2 (July 30th, 2020):
* [Change] Better handling of SSL/API errors in the `CurseClient`.
* [Fix] Opt-out from certificate validation when calling `CurseForge` and `Cloudflare`.
* [Fix] Supply `user-agent` header to make `Cloudflare` happy.

# KPSPDev: ReleaseTools v1.1 (October 29th, 2018):
* [Enhancement] Add `--github` argument to have the GitHub CHANGELOG links resolved.
* [Fix] Properly handle the Curse versions with no third component (e.g. "1.5").
* [Change] Rename argument SpaceDock's `--tag_extarct` to `--version_extract`.
* [Change] Introduce a package to handle the CHANGELOG markup.
* [Change] Major folder refactoring of the repo.

# KPSPDev: ReleaseTools v1.0 (July 19th, 2018):
* [Wiki] [New topic](https://github.com/ihsoft/KSPDev_ReleaseBuilder/wiki/Release-publishing-tools)
* [Enhancement] Add CurseForge publishing script.
* [Enhancement] Add Spacedock publishing script.
* [Enhancement] Add GitHub publishing script.

# KPSPDev: ReleaseBuilder v2.1 (August 15th, 2017):
* [Enhancement] Add new configuration parameter [RELEASE_NAME_FREE_FORMAT](https://github.com/ihsoft/KSPDev/wiki/ReleaseBuilder-Schema-1.1).
* [Enhancement] Support [auto-generated version components](https://msdn.microsoft.com/en-us/library/system.reflection.assemblyversionattribute%28v=vs.110%29.aspx#Anchor_6) for BUILD and REVISION.

# KPSPDev: ReleaseBuilder v2.0.0 (January 21st, 2017):
* [Change] All settings are simplified and renamed to give better context. JSON settings created
  for builder older than 1.2 won't work!
* [Change] Introduced JSON field `JSON_SCHEMA_VERSION` to help managing changes in JSON semantics
  in the future versions.
* [Change] Deprecated `POST_BUILD_COPY` setting. Such actions should be done via project's
  post-build events.
* [Change] Merge excutable and the builder class into one file for simplicity.
* [Enhancement] Do safety checks on the constructed paths: read paths must not be above project's
  root; write paths must not be above release folder path.
* [Enhancement] Major improvement of errors handling and reporting.
* [Enhancement] Added auto-detection of GitHub repository for `PROJECT_ROOT`.
* [Enhancement] Using Python built-in ZIP archiving capability. No need in external archiver
  anymore.
* [Enhancement] Implemented `-j` option that allows specifying arbitrary JSON settings file.
  When used in this mode working directory is assumed to be directory of the JSON file. Script
  doesn't need to be located in the building repository anymore.
* [Enhancement] Added support of macros in all paths.
* [Enhancement] Add more program arguments to better control builder behavior. Running wihtout any
  argument now shows help screen.


FormDataUtil.py
"""Provides helpers to deal with the multipart/form-data MIME types."""
import io
import json
import random
import string
import ntpath


def IdGenerator(size, chars=string.ascii_letters + string.digits):
  """Makes a unique random string of the requested size."""
  return ''.join(random.choice(chars) for _ in range(size))


def EncodeFormData(fields):
  """Encodes the provided arguments as the web form fields.

  The argument must be an array of objects that define the fields to be passed:
  - 'name': A required property than defines the name of the web form field.
  - 'data': The raw data to pass. If it's of type 'string', then it's passed as
    'text/plain'. Otherwise, it's serialized as JSON and passed as
    'application/json'.
  - 'filename': An optional property that designates a path to the binary file
    to tansfer. The 'data' field is ignored in this case, and the real data is
    read from the file. The data is passed as 'application/octet-stream', and
    server will receive the file name part of the path. The part can be
    relative, in which case it's counted realtive to the main module, or it can
    be absolute.

  Example:
    [
        { 'name': 'metadata', 'data': metadatObj },
        { 'name': 'file', 'filename': '/home/me/releases/MyMod_v1.0.zip' }
    ]

  @param fields: An array of the fields to encode.
  """
  boundry = '----WebKitFormBoundary' + IdGenerator(16)

  data_stream = io.BytesIO()
  for field in fields:
    filename = field.get('filename')
    if filename:
      with open(filename, 'rb') as f:
        data = f.read()
      content_type = 'application/octet-stream'
      filename = ntpath.basename(filename)
    else:
      data = field['data']
      if type(data) is not str:
        content_type = 'application/json'
        data = json.dumps(data)
      else:
        content_type = 'text/plain'
    _WriteFormData(data_stream, boundry,
                   field['name'], content_type, data, filename)

  data_stream.write(b'--')
  data_stream.write(boundry.encode())
  data_stream.write(b'--')
  data_stream.write(b'\r\n')

  headers = {
      'Content-Type': 'multipart/form-data; boundary=%s' % boundry,
      'Content-Length': str(data_stream.tell()),
  }
  return headers, data_stream.getvalue()


def _WriteFormData(stream, boundry, name, content_type, data, filename=None):
  """Helper method to write a single web form field."""
  stream.write(b'--')
  stream.write(boundry.encode())
  stream.write(b'\r\n')

  if filename:
    stream.write(('Content-Disposition: form-data; name="%s"; filename="%s"' %
                  (name, filename)).encode())
  else:
    stream.write(('Content-Disposition: form-data; name="%s"' %
                  (name)).encode())
  stream.write(b'\r\n')
  stream.write(('Content-Type: %s' % content_type).encode())
  stream.write(b'\r\n')
  stream.write(b'\r\n')
  stream.write(data)
  stream.write(b'\r\n')


PublishGitHub.py
# Public domain license.
# Author: igor.zavoychinskiy@gmail.com
# GitHub: https://github.com/ihsoft/KSPDev_ReleaseBuilder
# $version: 1
# $date: 07/17/2018

""" Script to publish releases to GitHub.

First of all you have to obtain the API token. It will authorize the requests.
Once you have it, define how you'd like the target release be created.

Example:

  $ PublishGitHub.py\
    --user=my_github_user\
    --repo=MyGitHubMod\
    --token=1111111111122222222222222333333333333333\
    --changelog=CHANGELOG.md\
    --archive=MyGitHubMod_v1.2.zip\
    --as_draft\
    --title=MyGitHubMod v{tag}

Note, the `--title` argument can be omitted. In this case the release will be
named as the archive file.

This script takes the release description from a CHANGELOG file. It assumes,
the topmost block of the lines till the first empty line is the description
of the latest release. If this is not the case, use the --changelog_breaker
parameter:

  --changelog_breaker=BreakHere

Every GitHub reelase must have a version tag. By default, it's extracted from
the archive name, using a default RegExp. If the archive's name is unusual,
then a custome RegExp needs to be specified:

  --tag_extract=MyName_v(.+?)

HINT. Passing all the arguments via the command line may be not convinient.
Not to mention the quotes and back slashes issues. To avoid all this burden,
simple put all the options into a text file and provide it to the script:

  $ PublishGitHub.py @my_params.txt

Don't bother about escaping in this file. Anything, placed in a line goes to
the parameter value *as-is*. So you don't need to escape backslashes,
whitespaces, quotes, etc.

When the script is properly configured and ran, it will present the extracted
portion of CHANGELOG and the other release settings, which will be applied to
the release. Review them and answer "y" if the look OK to have the release
created.

If you've provided `--as_draft` argument, then the release will be created but
not published, and there will be no version tags assigned to the trunk. This is
the recommended way to use the publish script. It allows creating a release
before committing and syncing the final release changes to the trunk. In order
to have the release published, go to GitHub and edit the draft.
"""
import argparse
import atexit
import os.path
import logging
import logging.config
import re
import sys
import textwrap

from clients import GitHubClient


LOGGER = logging.getLogger()


def main(argv):
  global LOGGER

  if os.path.isfile('logging.conf'):
    logging.config.fileConfig('logging.conf')
    atexit.register(_Shutdown)
    LOGGER = logging.getLogger('publishApp')
  else:
    print 'Log config not found: logging.conf'

  LOGGER.info('Staring new GitHub session...')

  parser = argparse.ArgumentParser(
      description='Publishes the release to a GitHub repository.',
      fromfile_prefix_chars='@',
      formatter_class=argparse.RawDescriptionHelpFormatter,
      epilog=textwrap.dedent('''
          Arguments can be provided via a file:
            %(prog)s @input.txt
      '''))
  parser.add_argument(
      '--user', action='store', metavar='<GitHub user>', required=True,
      help='''the user name on GitHub''')
  parser.add_argument(
      '--repo', action='store', metavar='<GitHub repo>', required=True,
      help='''the repository name on GitHub''')
  parser.add_argument(
      '--token', action='store', metavar='<Personal Access Token>',
      required=True,
      help='''the token to authorize in API. To obtain this token go to the
          repository on GitHub, choose: "Settings/Developer settings/Personal
          access tokens". Generate a token with permision "repo".''')
  parser.add_argument(
      '--changelog', action='store', metavar='<file path>', required=True,
      help='''the file to get the release description from. The top lines till
          the first empty line are taken. The description is expected to use
          the 'markdown' syntax.''')
  parser.add_argument(
      '--archive', action='store', metavar='<file path>', required=True,
      help='''the archive file to publish.''')
  parser.add_argument(
      '--tag_extract', action='store', metavar='<regexp>',
      default='.+?_v(.+?)\\.zip',
      help='''the RegExp to extract the version tag from the archive name.
          [Default: %(default)s]''')
  parser.add_argument(
      '--changelog_breaker', action='store', metavar='<regexp>',
      default=r'^\s*$',
      help='''the RegExp to detect the end of the release description in the
          CHANGELOG. This expression is applied per the file line.
          [Default: %(default)s]''')
  parser.add_argument(
      '--title', action='store', metavar='<pattern>',
      help='''the pattern to build the release name. Use placeholder {tag} for
          the version tag. If omitted, then the file name is used as the one.
          Example: "NewName_{tag}".''')
  parser.add_argument(
      '--as_draft', action='store_true',
      help='''forces the created release to be draft''')
  opts = vars(parser.parse_args(argv[1:]))

  user = opts['user']
  repo = opts['repo']

  # Init GitHub client.
  GitHubClient.API_TOKEN = opts['token']

  desc = _ExtractDescription(opts['changelog'], opts['changelog_breaker'])
  filename = opts['archive']
  as_draft = opts['as_draft']

  parts = re.findall(opts['tag_extract'], os.path.basename(filename))
  if len(parts) != 1:
    print 'ERROR: cannot extract version tag from file name: %s' % filename
    exit(-1)
  version_tag = parts[0]

  if opts['title']:
    title = opts['title'].format(tag=version_tag)
  else:
    title = os.path.splitext(os.path.basename(filename))[0]

  if not os.path.isfile(filename):
    print 'ERROR: Cannot find archive: %s' % filename
    exit(-1)

  # Verify the user's choice...
  print '======> BEGIN CHANGELOG:'
  print desc
  print '======> END CHANGELOG:'
  print
  print 'GitHub repo: %s/%s' % (user, repo)
  print 'Upload file:', os.path.abspath(filename)
  print 'Name release as:', title
  print 'Release status:', ('DRAFT' if as_draft else 'PUBLISHED')
  sys.stdout.write('\nContinue? [y/N]: ')
  choice = raw_input().lower()
  if choice != 'y' and choice != 'Y':
    LOGGER.info('ABORTED BY USER!')
    print 'ABORTED!'
    exit(-1)

  print 'Publishing the release...'
  GitHubClient.CreateRelease(
      user, repo, version_tag, title, desc,
      files=[{'filename': filename }],
      draft=opts['as_draft'])
  print 'DONE!'
  if opts['as_draft']:
    print 'You must publish the release to make it visible!'


def _Shutdown():
  """Finilizes the logging system."""
  LOGGER.info('Ending GitHub session...')
  logging.shutdown()


def _ExtractDescription(changelog_file, breaker_re):
  """Helper method to extract the meaningful part of the release changelog."""
  with open(changelog_file, 'r') as f:
    lines= f.readlines()
  changelog = ''
  for line in lines:
    # Ignore any trailing empty lines.
    if not changelog and not line.strip():
      continue
    # Stop at the breaker.
    if re.match(breaker_re, line.strip()):
      break
    changelog += line
  return changelog.strip()


try:
  main(sys.argv)
except Exception, ex:
  LOGGER.exception(ex)


PublishSpacedock.py
# Public domain license.
# Author: igor.zavoychinskiy@gmail.com
# GitHub: https://github.com/ihsoft/KSPDev_ReleaseBuilder
# $version: 4
# $date: 10/28/2018

""" Script to publish releases to Spacedock.

Example:

  $ PublishSpacedock.py\
    --project=1906
    --changelog=CHANGELOG.md
    --versions=^1\.4\.4$
    --archive=./test_v1.5.zip

This script takes the release description from a CHANGELOG file. It assumes,
the topmost block of the lines till the first empty line is the description
of the latest release. If this is not the case, use the --changelog_breaker
parameter:

  --changelog_breaker=BreakHere

The script needs to know the mod version to report to Spacedock. By default,
it takes it from the release arhcive name, assuming it has format:
'*_v<version>.zip'. When it's different, you need to specify the extraction
RegExp:

  --version_extract=.+?_v(.+?)\\.zip

HINT. Passing all the arguments via the command line may be not convinient.
Not to mention the quotes and back slashes issues. To avoid all this burden,
simple put all the options into a text file and provide it to the script:

  $ PublishSpacedock.py @my_params.txt

Don't bother about escaping in this file. Anything, placed in a line goes to
the parameter value *as-is*. So you don't need to escape backslashes,
whitespaces, quotes, etc.

Example of the args file (you may copy it "as-is"):

  --project=123456
  --login=foo
  --changelog=CHANGELOG.md
  --version=1\.4\.
  --archive=./test_v1.5.zip

When the script is properly configured and ran, it will present the extracted
portion of CHANGELOG and the other release settings, which will be applied to
the project. Review them *CAREFULLY* before answering "y". The new versions
will be published right away!
"""
import atexit
import argparse
import getpass
import logging
import logging.config
import os.path
import re
import sys
import textwrap

from clients import SpacedockClient
from utils import ChangelogUtils


LOGGER = logging.getLogger()


def main(argv):
  global LOGGER

  if os.path.isfile('logging.conf'):
    logging.config.fileConfig('logging.conf')
    atexit.register(_Shutdown)
    LOGGER = logging.getLogger('publishApp')
  else:
    print 'Log config not found: logging.conf'

  LOGGER.info('Staring new SpaceDock session...')

  parser = argparse.ArgumentParser(
      description='Publishes the release to a Spacedock mod.',
      fromfile_prefix_chars='@',
      formatter_class=argparse.RawDescriptionHelpFormatter,
      epilog=textwrap.dedent('''
          Arguments can be provided via a file:
            %(prog)s @input.txt
      '''))
  parser.add_argument(
      '--project', action='store', metavar='<project ID>', required=True,
      help='''the ID of the project to publish to. To get it, go to the mod
          overview on Spacedock and extract the number from the URL:
          /mod/<project ID>/...''')
  parser.add_argument(
      '--changelog', action='store', metavar='<file path>', required=True,
      help='''the file to get the release description from. The top lines till
          the first empty line are taken. The description is expected to use
          the 'markdown' syntax.''')
  parser.add_argument(
      '--archive', action='store', metavar='<file path>', required=True,
      help='''the archive file to publish.''')
  parser.add_argument(
      '--ksp_version', action='store', metavar='<"latest" | regexp>',
      default='latest',
      help='''the RegExp pattern to match the target KSP version. If set to the
          keyword "latest", then the script will use the maximum version,
          known to Spacedock. [Default: %(default)s]''')
  parser.add_argument(
      '--changelog_breaker', action='store', metavar='<regexp>',
      default=r'^\s*$',
      help='''the RegExp to detect the end of the release description in the
          CHANGELOG. This expression is applied per the file line.
          [Default: %(default)s]''')
  parser.add_argument(
      '--version_extract', action='store', metavar='<regexp>',
      default='.+?_v(.+?)\\.zip',
      help='''the RegExp to extract the version tag from the archive name.
          [Default: %(default)s]''')
  parser.add_argument(
      '--login', action='store', metavar='<SD login>',
      help='''the login for the Spacedock account. If not set, then it will be
          asked in the command line.''')
  parser.add_argument(
      '--pass', action='store', metavar='<SD password>',
      help='''the password for the Spacedock account. If not set, then it will
          be asked in the command line.''')
  opts = vars(parser.parse_args(argv[1:]))

  mod_id = opts['project']

  versions_re = opts['ksp_version']
  if versions_re != 'latest':
    all_versions = map(
        lambda x: x['name'],
        SpacedockClient.GetKSPVersions(pattern=versions_re))
    if not all_versions:
      print 'ERROR: No versions found for RegExp: %s' % versions_re
      exit(-1)
    if len(all_versions) > 1:
      print 'ERROR: Multiple versions matched RegExp: %s' % all_versions
      exit(-1)
    game_version = all_versions[0]
  else:
    all_versions = sorted(
        SpacedockClient.GetKSPVersions(), key=lambda x: x['id'], reverse=True)
    if not all_versions:
      print 'ERROR: No versions found!'
      exit(-1)
    game_version = all_versions[0]['name']

  desc = ChangelogUtils.ExtractDescription(
      opts['changelog'], opts['changelog_breaker'])
  filename = opts['archive']

  parts = re.findall(opts['version_extract'], os.path.basename(filename))
  if len(parts) != 1:
    print 'ERROR: cannot extract version tag from file name: %s' % filename
    exit(-1)
  mod_version = parts[0]

  if not os.path.isfile(filename):
    print 'ERROR: Cannot find archive: %s' % filename
    exit(-1)

  try:
    mod_details = SpacedockClient.GetModDetails(mod_id)
  except SpacedockClient.AuthorizationRequiredError, ex:
    # The unpublished mods are not available via API (see issue #186).
    mod_details = { 'id': mod_id, 'name': '<UNPUBLISHED>' }

  # Verify the user's choice...
  print '======> BEGIN CHANGELOG:'
  print desc
  print '======> END CHANGELOG:'
  print
  print 'Mod name: %s (#%s)' % (mod_details['name'], mod_details['id'])
  print 'Upload file:', os.path.abspath(filename)
  print 'Mod version tag:', mod_version
  print 'Add for version:', game_version

  sys.stdout.write('\nContinue? [y/N]: ')
  choice = raw_input().lower()
  if choice != 'y' and choice != 'Y':
    LOGGER.info('ABORTED BY USER!')
    print 'ABORTED!'
    exit(-1)

  print 'Publishing the release...'

  # Init Spacedock client.
  login = opts['login']
  if not login:
    sys.stdout.write('Spacedock login: ')
    login = raw_input().lower()
  else:
    print 'Spacedock login:', login
  SpacedockClient.API_LOGIN = login

  password = opts['pass']
  if not password:
    password = getpass.getpass('Spacedock password: ')
  else:
    print 'Spacedock password: <PROVIDED>'
  SpacedockClient.API_PASS = password

  SpacedockClient.UploadFile(mod_id, filename, desc, mod_version, game_version)
  print 'DONE!'
  print 'The new version is now availabe and the followers are notified!'


def _Shutdown():
  """Finilizes the logging system."""
  LOGGER.info('Ending SpaceDock session...')
  logging.shutdown()


main(sys.argv)


PublishCurseForge.py
# Public domain license.
# Author: igor.zavoychinskiy@gmail.com
# GitHub: https://github.com/ihsoft/KSPDev_ReleaseBuilder
# $version: 3
# $date: 10/28/2018

""" Script to publish releases to Kerbal CurseForge.

First of all you have to obtain the API token. It will authorize the requests.
Once you have it, define how you'd like the target release be created.

Example:

  $ PublishCurseForge.py\
    --project=123456
    --token=11111111-2222-3333-4444-555555555555
    --changelog=CHANGELOG.md
    --versions=1\.4\.
    --archive=./test_v1.5.zip

This script takes the release description from a CHANGELOG file. It assumes,
the topmost block of the lines till the first empty line is the description
of the latest release. If this is not the case, use the --changelog_breaker
parameter:

  --changelog_breaker=BreakHere

By default, the script will name the reelase the same way as the archive.
However, you may request to change the behavior. For this, you need to specify
how to extract the versions specific tag from the archive file name. Then, you
need to provide a new pattern for the relase name.

  --tag_extract=.+?_(v.+?)\\.zip
  --title=NewName_{tag}\

Note, that "tag_extract" has a default value, which extracts the archive name
version suffix. E.g. if the name was "my_mod_v1.5.zip", then the tag will be
"v1.5".

HINT. Passing all the arguments via the command line may be not convinient.
Not to mention the quotes and back slashes issues. To avoid all this burden,
simple put all the options into a text file and provide it to the script:

  $ PublishCurseForge.py @my_params.txt

Don't bother about escaping in this file. Anything, placed in a line goes to
the parameter value *as-is*. So you don't need to escape backslashes,
whitespaces, quotes, etc.

Example of the args file (you may copy it "as-is"):

  --project=123456
  --token=11111111-2222-3333-4444-555555555555
  --changelog=CHANGELOG.md
  --versions=1\.4\.
  --archive=./test_v1.5.zip
  --title=MyMod "Awesome!" {tag}

When the script is properly configured and ran, it will present the extracted
portion of CHANGELOG and the other release settings, which will be applied to
the project. Review them *CAREFULLY* before answering "y". CurseForge doesn't
allow deleting the relases. Regardless to how bad was your attempt, it will
stay in the history forever (but you can archive it).
"""
import argparse
import atexit
import os.path
import logging
import logging.config
import re
import sys
import textwrap

from clients import CurseForgeClient
from utils import ChangelogUtils


LOGGER = logging.getLogger()


def main(argv):
  global LOGGER

  if os.path.isfile('logging.conf'):
    logging.config.fileConfig('logging.conf')
    atexit.register(_Shutdown)
    LOGGER = logging.getLogger('publishApp')
  else:
    print 'Log config not found: logging.conf'

  LOGGER.info('Staring new CurseForge session...')

  parser = argparse.ArgumentParser(
      description='Publishes the release to a Kerbal CurseForge project.',
      fromfile_prefix_chars='@',
      formatter_class=argparse.RawDescriptionHelpFormatter,
      epilog=textwrap.dedent('''
          Arguments can be provided via a file:
            %(prog)s @input.txt
      '''))
  parser.add_argument(
      '--project', action='store', metavar='<project ID>', required=True,
      help='''the ID of the project to publish to. To get it, go to the project
          overview on CurseForge.''')
  parser.add_argument(
      '--token', action='store', metavar='<API token>', required=True,
      help='''the token to authorize in API. To obtain this token go to the
          mod's profile on Forge, choose: "Account/Preferences/My API
          Tokens"''')
  parser.add_argument(
      '--changelog', action='store', metavar='<file path>', required=True,
      help='''the file to get the release description from. The top lines till
          the first empty line are taken. The description is expected to use
          the 'markdown' syntax.''')
  parser.add_argument(
      '--archive', action='store', metavar='<file path>', required=True,
      help='''the archive file to publish.''')
  parser.add_argument(
      '--title', action='store', metavar='<pattern>',
      help='''the pattern to build the release name. Use placeholder {tag} for
          the version tag. If omitted, then the file name is used as the one.
          Example: "NewName_{tag}".''')
  parser.add_argument(
      '--versions', action='store',
      metavar='<"latest" | "latest_all_builds" | regexp>',
      default='latest_all_builds',
      help='''the pattern to match the target KSP versions. If "latest" is
         specified, then a single version is used - the most recent one. If
         "latest_all_builds" is specified, then all builds of the latest
         version are used (e.g. "1.4.*"). [Default: %(default)s]''')
  parser.add_argument(
      '--changelog_breaker', action='store', metavar='<regexp>',
      default=r'^\s*$',
      help='''the RegExp to detect the end of the release description in the
          CHANGELOG. This expression is applied per the file line.
          [Default: %(default)s]''')
  parser.add_argument(
      '--tag_extract', action='store', metavar='<regexp>',
      default='.+?_(v.+?)\\.zip',
      help='''the RegExp to extract the version tag from the archive name.
          [Default: %(default)s]''')
  opts = vars(parser.parse_args(argv[1:]))

  project_id = opts['project']

  # Init CurseForge client.
  CurseForgeClient.API_TOKEN = opts['token']

  versions_re = opts['versions']
  if versions_re == 'latest' or versions_re == 'latest_all_builds':
    game_versions = sorted(
        CurseForgeClient.GetKSPVersions(), key=lambda x: x['id'], reverse=True)
    if not game_versions:
      print 'ERROR: No versions found!'
      exit(-1)
    game_versions = map(lambda x: x['name'], game_versions)
    if versions_re == 'latest':
      game_versions = game_versions[0:1]
    else:
      prefix = re.match(r'^(.+?\..+?)(\..+)?$', game_versions[0]).group(1)
      game_versions = filter(lambda x: x.startswith(prefix), game_versions)
  else:
    game_versions = map(
        lambda x: x['name'],
        CurseForgeClient.GetKSPVersions(pattern=versions_re))
    if not game_versions:
      print 'ERROR: No versions found for RegExp: %s' % versions_re
      exit(-1)

  desc = ChangelogUtils.ExtractDescription(
      opts['changelog'], opts['changelog_breaker'])
  filename = opts['archive']

  if opts['title']:
    if not opts['tag_extract']:
      print ('ERROR: When title rewrite is requested, the "tag_extract"'
             ' parameter is required')
      exit(-1)
    parts = re.findall(opts['tag_extract'], os.path.basename(filename))
    if len(parts) != 1:
      print 'ERROR: cannot extract version tag from file name: %s' % filename
      exit(-1)
    title = opts['title'].format(tag=parts[0])
  else:
    title = os.path.splitext(os.path.basename(filename))[0]


  if not os.path.isfile(filename):
    print 'ERROR: Cannot find archive: %s' % filename
    exit(-1)

  project_details = CurseForgeClient.GetProjectDetails(project_id)

  # Verify the user's choice...
  print '======> BEGIN CHANGELOG:'
  print desc
  print '======> END CHANGELOG:'
  print
  print 'Project name: %s (#%s)' % (project_details['title'], project_id)
  print 'Upload file:', os.path.abspath(filename)
  print 'Name release as:', title
  print 'Add for versions:', ', '.join(game_versions)
  sys.stdout.write('\nContinue? [y/N]: ')
  choice = raw_input().lower()
  if choice != 'y' and choice != 'Y':
    LOGGER.info('ABORTED BY USER!')
    print 'ABORTED!'
    exit(-1)

  print 'Publishing the release...'
  CurseForgeClient.UploadFile(
      project_id, filename, desc, game_versions,
      title=title, release_type='release')
  print 'DONE!'
  print 'Watch for the verification status on CurseForge.'


def _Shutdown():
  """Finilizes the logging system."""
  LOGGER.info('Ending CurseForge session...')
  logging.shutdown()


main(sys.argv)
